include/boost/beast/core/detail/any_buffers.hpp

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright (c) 2016-2021 Vinnie Falco (vinnie dot falco at gmail dot com)
+// Copyright (c) 2021 Richard Hodges (hodges.r@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef BOOST_BEAST_CORE_DETAIL_ANY_BUFFERS_HPP
+#define BOOST_BEAST_CORE_DETAIL_ANY_BUFFERS_HPP
+
+#include <boost/beast/core/detail/config.hpp>
+
+#include <boost/asio/buffer.hpp>
+
+#include <atomic>
+#include <cstddef>
+#include <memory>
+#include <type_traits>
+
+namespace boost {
+namespace beast {
+namespace detail {
+
+template<bool IsConst>
+class any_buffers
+{
+public:
+    using value_type =
+        typename std::conditional<
+            IsConst,
+            net::const_buffer,
+            net::mutable_buffer>::type;
+
+private:
+    struct impl_base
+    {
+        std::atomic<std::size_t> refs = { 1 };
+        std::size_t const size;
+        value_type const* buffers;
+
+        std::size_t
+        addref() noexcept;
+
+        std::size_t
+        release() noexcept;
+
+        virtual ~impl_base() = default;
+
+    private:
+        virtual
+        void
+        destroy() noexcept = 0;
+    };
+
+    template<
+        class Buffers,
+        class Alloc = std::allocator<void>>
+    struct impl : impl_base
+    {
+        impl(
+            Buffers const& b,
+            Alloc const& a)
+        : b_(b)
+        , a_(a)
+        {
+        }
+
+        Buffers b_;
+        Alloc   a_;
+
+    private:
+        void
+        destroy() noexcept override;
+    };
+
+    impl_base* impl_;
+
+    template<
+        class Buffers,
+        class Alloc = std::allocator<void>, typename std::enable_if<
+            !std::is_base_of<any_buffers, Buffers>::value >::type* = nullptr>
+    explicit
+    any_buffers(
+        Buffers const& b,
+        Alloc const& a = {})
+    : impl_(construct(b, a))
+    {
+    }
+
+    any_buffers(any_buffers const& r);
+
+    any_buffers(any_buffers && r) noexcept;
+
+    any_buffers&
+    operator=(any_buffers const& r);
+
+    any_buffers&
+    operator=(any_buffers && r);
```

Comment:
I don't think assignment should be supported

Reply:
now you mention it, that probably makes sense